chronos/ratelimit.nim

```diff
@@ -0,0 +1,101 @@
+#               Chronos Rate Limiter
+#            (c) Copyright 2022-Present
+#         Status Research & Development GmbH
+#
+#              Licensed under either of
+#  Apache License, version 2.0, (LICENSE-APACHEv2)
+#              MIT license (LICENSE-MIT)
+
+{.push raises: [Defect].}
+
+import ../chronos
+import timer
+
+export timer
+
+type
+  RateCounter* = object
+    budget: int64
+    budgetCap: int64
+    lastUpdate: Moment
+
+  RateCounterRef* = ref RateCounter
+
+  RateCost* = distinct int64
+
+proc update(rc: var RateCounter) =
+  let
+    currentTime = Moment.now()
+    timeDelta = nanoseconds(currentTime - rc.lastUpdate)
+
+  rc.lastUpdate = currentTime
+  rc.budget = min(rc.budgetCap, rc.budget + timeDelta)
+
+proc operationsPerSecond*(ops: float): RateCost =
+  ## Create a RateCost from an allowed number of 
+  ## operations per second
+  const budgetPerSecond = 1.seconds.nanoseconds.float
+
+  RateCost(budgetPerSecond / ops)
+
+proc timeBudgetPerSecond*(time, budget: Duration): RateCost =
+  ## Create a RateCost from a time budget per second.
+  runnableExamples:
+    let start = Moment.now()
+    # expensive computation
+    let cost = timeBudgetPerSecond(
+      Moment.now() - start,
+      300.milliseconds # Allow 300 milliseconds of
+                       # computation per second
+    )
+  let
+    timeAsUs = time.nanoseconds.float
+    budgetAsUs = budget.nanoseconds.float
+    budgetUsed = timeAsUs / budgetAsUs
+    allowedPerSeconds = 1.float / budgetUsed
+  operationsPerSecond(allowedPerSeconds)
+
+proc tryConsume*(rc: var RateCounter, cost: RateCost): bool =
+  ## If there is still budget left, remove cost from the budget
+  ## Otherwise, return false
+  rc.update()
+  if rc.budget >= 0:
+    rc.budget -= int64(cost)
+    true
+  else:
+    false
+
+proc consume*(rc: var RateCounter, cost: RateCost): Future[void] =
+  ## Wait until some budget is available, then substract cost
+  ## from it
+
+  # Manual async because of var argument
+  rc.update()
+  rc.budget -= int64(cost)
+
+  if rc.budget >= 0:
+    result = newFuture[void]("RateCounter.consume")
+    result.complete()
+    return result
+  else:
+    return sleepAsync(nanoseconds(-rc.budget))
+
+proc init*(T: typedesc[RateCounter], cap: Duration): T =
```

Comment:
Hm, I think we'll benefit from using a more standard setup of naming and initialization here - the [ones](https://github.com/osrg/openvswitch/blob/master/lib/token-bucket.c#L33) I've seen take a separate rate and cap (so that I can create a 100 tokens/sec bucket with a burst capacity of 1000, for example) - it's not obvious to me how to create such a bucket from these parameters.

Instead of "operations" I'd also go with "tokens" - there's a lot of [literature](http://intronetworks.cs.luc.edu/current/html/tokenbucket.html#token-bucket-definition) out there describing this particular algorithm so it's good if it's instantly recognizable - it should be documented with the name of the algo as well.

Reply:
In nimbus for example, we'll likely want to switch to having [two of these buckets](http://intronetworks.cs.luc.edu/current/html/tokenbucket.html#multiple-token-buckets): a smaller bursty one and a global more "even" one - when other clients ask for blocks, we want each peer to get an allocation but also put a global limit on how much bandwidth we  assign to providing blocks.

That said, what the "tokens" are is arbitrary - it's usually bytes, but sometimes packets or requests - there are arguments for each, depending on what's costly.

Reply:
Finally, I've been thinking about a manual replenish function as well: instead of a rate / sec, in some cases I might want to replenish manually: the loose idea I had was that we should not spend more on block uploads than we spend on gossip - this would mean that every time we send gossip, we'd replenish the block bucket up to its cap, so that max block bandwidth is expressed in terms of gossip bandwidth - it's a loose idea so far, but it does solve one problem which is how to "size" the buckets naturally.

Reply:
I thought a bit about it, and couldn't find a use case wherever scaling the "costs" instead of the fill rate didn't work

Hence my idea here to simplify the api to just burst capacity / cost, which is I think easier to reason about

But indeed, to handle manual replenish, we might have to expose the fill rate 🤔